jib-core/src/main/java/com/google/cloud/tools/jib/registry/ManifestChecker.java

```diff
@@ -0,0 +1,190 @@
+            + "/manifests/"
+            + imageDescriptor);
+  }
+
+  @Override
+  public String getHttpMethod() {
+    return HttpMethods.GET;
+  }
+
+  @Override
+  public String getActionDescription() {
+    return "pull image manifest for "
+        + registryEndpointRequestProperties.getServerUrl()
+        + "/"
+        + registryEndpointRequestProperties.getImageName()
+        + ":"
+        + imageDescriptor;
+  }
+
+  /**
+   * Instantiates a {@link ManifestTemplate} from a JSON string. This checks the {@code
+   * schemaVersion} field of the JSON to determine which manifest version to use.
+   */
+  private T getManifestTemplateFromJson(String jsonString)
+      throws IOException, UnknownManifestFormatException {
+    ObjectNode node = new ObjectMapper().readValue(jsonString, ObjectNode.class);
+    if (!node.has("schemaVersion")) {
+      throw new UnknownManifestFormatException("Cannot find field 'schemaVersion' in manifest");
+    }
+
+    int schemaVersion = node.get("schemaVersion").asInt(-1);
+    if (schemaVersion == -1) {
+      throw new UnknownManifestFormatException("`schemaVersion` field is not an integer");
+    }
+
+    if (schemaVersion == 1) {
+      return manifestTemplateClass.cast(
+          JsonTemplateMapper.readJson(jsonString, V21ManifestTemplate.class));
+    }
+    if (schemaVersion == 2) {
+      // 'schemaVersion' of 2 can be either Docker V2.2 or OCI.
+      String mediaType = node.get("mediaType").asText();
+      if (V22ManifestTemplate.MANIFEST_MEDIA_TYPE.equals(mediaType)) {
+        return manifestTemplateClass.cast(
+            JsonTemplateMapper.readJson(jsonString, V22ManifestTemplate.class));
+      }
+      if (OciManifestTemplate.MANIFEST_MEDIA_TYPE.equals(mediaType)) {
+        return manifestTemplateClass.cast(
+            JsonTemplateMapper.readJson(jsonString, OciManifestTemplate.class));
+      }
+      if (V22ManifestListTemplate.MANIFEST_MEDIA_TYPE.equals(mediaType)) {
+        return manifestTemplateClass.cast(
+            JsonTemplateMapper.readJson(jsonString, V22ManifestListTemplate.class));
+      }
+      throw new UnknownManifestFormatException("Unknown mediaType: " + mediaType);
+    }
+    throw new UnknownManifestFormatException(
+        "Unknown schemaVersion: " + schemaVersion + " - only 1 and 2 are supported");
+  }
+
+  @Override
+  public Optional<ManifestAndDigest<T>> handleHttpResponseException(
+      ResponseException responseException) throws ResponseException {
+    if (responseException.getStatusCode() != HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+      throw responseException;
+    }
+
+    // Finds a MANIFEST_BLOB_UNKNOWN error response code.
+    if (responseException.getContent() == null) {
+      return Optional.empty();
+    }
+
+    ErrorCodes errorCode = ErrorResponseUtil.getErrorCode(responseException);
+    if (errorCode == ErrorCodes.MANIFEST_UNKNOWN) {
+      return Optional.empty();
+    }
```

Comment:
Could you explain these two checks (`responseException.getContext()` and `ErrorCodes.MANIFEST_UNKNOWN`)? I am asking because we've seen different registries have different behaviors that are not really conformant to the Registry API standard, and would like to know if these checks are specific to a certain registry implementation.

Reply:
I have followed the same pattern done in `BlobChecker`.

Reply:
Alright, I will look the Registry API specification later for this part and will probably take care of this one way or another in a follow-up PR. But I think at least this works and doesn't actually break things, since this is just an extra manifest check; the worst case would be getting an error later. So I think we are good to go.

Reply:
This looks good.